File: n_queens.java
import java.util.Scanner;

public class n_queens {
    private int n;
    private char[][] board;

    public n_queens(int n) {
        this.n = n;
        board = new char[n][n];

        // Initialize board with empty spaces
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                board[i][j] = '.';
    }

    // Function to check if a queen can be placed at (row, col)
    private boolean isSafe(int row, int col) {
        // Check vertical column
        for (int i = 0; i < row; i++)
            if (board[i][col] == 'Q')
                return false;

        // Check upper-left diagonal
        for (int i = row, j = col; i >= 0 && j >= 0; i--, j--)
            if (board[i][j] == 'Q')
                return false;

        // Check upper-right diagonal
        for (int i = row, j = col; i >= 0 && j < n; i--, j++)
            if (board[i][j] == 'Q')
                return false;

        return true; // Safe to place queen
    }

    // Backtracking function to place queens row by row
    private boolean solve(int row) {
        if (row == n) {
            printBoard();
            return true; // Solution found
        }

        for (int col = 0; col < n; col++) {
            if (isSafe(row, col)) { // Check if the move is valid
                board[row][col] = 'Q'; // Place the queen
                if (solve(row + 1)) // Recur for next row
                    return true;
                board[row][col] = '.'; // Backtrack
            }
        }
        return false; // No valid placement found for this row
    }

    public void solveNQueens() {
        if (!solve(0))
            System.out.println("No solution exists!");
    }

    // Function to print the board
    private void printBoard() {
        for (char[] row : board) {
            for (char cell : row)
                System.out.print(cell + " ");
            System.out.println();
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter value of N: ");
        int n = scanner.nextInt();
        scanner.close();

        n_queens nQueens = new n_queens(n);
        nQueens.solveNQueens();
    }
}

File: nqueens.py
# Function to check if placing a queen at (row, col) is safe
def is_safe(board, row, col, n):
    # Check for queen in the same column in previous rows
    for i in range(row):
        if board[i][col] == 1:
            return False
        # Check the left upper diagonal
        if col - row + i >= 0 and board[i][col - row + i] == 1:
            return False
        # Check the right upper diagonal
        if col + row - i < n and board[i][col + row - i] == 1:
            return False
    # If no conflicts, it's safe to place a queen
    return True

# Recursive function to solve the N-Queens problem
def solve(board, row, n):
    # Base case: If all queens are placed
    if row == n:
        # Print the board configuration
        for r in board:
            # Print 'Q' for queen and '.' for empty cell
            print(" ".join(['Q' if x == 1 else '.' for x in r]))
        print()  # Empty line between solutions
        return

    # Try placing queen in all columns of the current row
    for col in range(n):
        # Check if it's safe to place a queen at (row, col)
        if is_safe(board, row, col, n):
            # Place the queen
            board[row][col] = 1
            # Recur to place the rest of the queens
            solve(board, row + 1, n)
            # Backtrack: Remove the queen for trying next possibility
            board[row][col] = 0

# Main part of the program

# Ask the user to input the value of N
try:
    n = int(input("Enter the value of N for the N-Queens problem: "))
    
    # Validate that N is a positive number
    if n <= 0:
        print("Please enter a positive integer greater than 0.")
    else:
        # Initialize an empty NxN chessboard with all zeros
        board = [[0]*n for _ in range(n)]
        # Call the solve function starting from the first row
        solve(board, 0, n)

# Handle the case where input is not an integer
except ValueError:
    print("Invalid input. Please enter a valid integer.")
